drivers/volume/portworx/portworx.go

```diff
@@ -1778,6 +1781,287 @@ func (d *portworx) ValidatePureVolumesNoReplicaSets(volumeName string, params ma
+				foundDevices[currentEntry.WWID] = *currentEntry
+				currentEntry = nil
+			}
+		}
+	}
+	if currentEntry != nil {
+		foundDevices[currentEntry.WWID] = *currentEntry
+	}
+
+	return foundDevices, nil
+}
+
+// collectLocalNodeInfo interrogates dmsetup and lsblk to get a comprehensive list of mapper devices, their single paths,
+// and checks for common error conditions such as devices missing single paths or only appearing in dmsetup/lsblk
+func (d *portworx) collectLocalNodeInfo(n node.Node) (map[string]pureLocalPathEntry, error) {
+	// Run `dmsetup ls` to get all the known mapper devices
+	out, err := d.nodeDriver.RunCommandWithNoRetry(n, "dmsetup ls", node.ConnectionOpts{
+		Timeout:         crashDriverTimeout,
+		TimeBeforeRetry: defaultRetryInterval,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to check dmsetup ls output on node %s, Err: %v", n.MgmtIp, err)
+	}
+	dmsetupFoundMappers := []string{}
+	for _, line := range strings.Split(out, "\n") {
+		if !strings.Contains(line, "24a937") {
+			continue
+		}
+		mapperName := strings.Split(line, "\t")[0]
+		dmsetupFoundMappers = append(dmsetupFoundMappers, mapperName)
+	}
+
+	// Then run `lsblk` to get the WWN and size of each device. Flags:
+	// * --inverse: show parents before children (instead of the default output that has single paths as the top level)
+	// * --ascii: show ASCII characters only (no unicode, normally it displays the fancy tree characters, this makes it only show |- and `-)
+	// * --noheadings: don't show the header line
+	// * -o NAME,SIZE: only show the name and size columns
+	// * -b: show size in bytes instead of human-readable with a suffix
+	out, err = d.nodeDriver.RunCommandWithNoRetry(n, "lsblk --inverse --ascii --noheadings -o NAME,SIZE -b", node.ConnectionOpts{
+		Timeout:         crashDriverTimeout,
+		TimeBeforeRetry: defaultRetryInterval,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to check 'lsblk --inverse --ascii --noheadings -o NAME,SIZE' output on node %s, Err: %v", n.MgmtIp, err)
+	}
+
+	lsblkParsed, err := parseLsblkOutput(out)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse lsblk output on node %s, Err: %v", n.MgmtIp, err)
+	}
+
+	// Also, raise an error if there is a mismatch in the devices in the two outputs (e.g. if there is a mapper
+	// in dmsetup but not lsblk, something is very wrong and we didn't clean up right)
+	for _, mapper := range dmsetupFoundMappers {
+		lsblkEntry, exists := lsblkParsed[mapper]
+		if !exists {
+			return nil, fmt.Errorf("found mapper %s in dmsetup but not lsblk on node %s, inconsistent disk state (we didn't clean something up right?)", mapper, n.MgmtIp)
+		}
+		// Ensure there are no WWIDs with empty single paths, as that is a malformed device
+		if len(lsblkEntry.SinglePaths) == 0 {
+			return nil, fmt.Errorf("found mapper %s with no single paths on node %s, this should not happen (we didn't clean something up right?)", mapper, n.MgmtIp)
+		}
+	}
+
```

Comment:
Add a check for `len(lsblkParsed) == len(dmsetupFoundMappers)`, if for some reason dmsetup did not find a mapper path but lsblk found one.

Reply:
Actually this might not work since in `parseLsblkOutput` you are parsing all devices, even if they are not multipath. May be a keep count of mapper devices found in `lsblk` output and compare it with that.